electrumsv/wallet_support/late_header_worker.py

```diff
@@ -115,94 +133,123 @@ async def _process_one_item(db_functions_async: AsynchronousFunctions,
-        # As we have confirmed that at this point the header is not present, monitor it.
-        assert tsc_proof.transaction_hash is not None
-        state.block_transactions[tsc_proof.block_hash].add(tsc_proof.transaction_hash)
-    else:
-        # Process all deferred transaction processing for the header's block.
-        assert tsc_proof.transaction_hash is not None
+        # We have confirmed that at this point the header is not present, monitor it.
         state.block_transactions[tsc_proof.block_hash].add(tsc_proof.transaction_hash)
-        await _process_block_transactions(db_functions_async, state, header, chain)
+        return None
+
+    confirmations = get_confirmations_or_handle_orphan(chain, header)
+    unverified_entries = await db_functions_async.read_transaction_proof_data_async(
+        [tsc_proof.transaction_hash])
+    tx_hash, flags, tx_block_hash, proof_data = unverified_entries[0]
+    try:
+        assert tx_block_hash is not None
+        assert proof_data is not None
+        await _process_one_merkle_proof(db_functions_async, state, tx_block_hash, flags,
+            tx_block_hash, proof_data, confirmations, header)
+    except MerkleProofProcessingError as e:
+        if e.error_type == MerkleProofErrorType.BAD_SERVER:
+            raise  # crash  # TODO(1.4.0) Handle all error types
 
 
 async def _process_header(db_functions_async: AsynchronousFunctions,
         state: LateHeaderWorkerState, header: Header, chain: Chain) -> None:
+    """Process merkle proofs in a batch-wise fashion for the given block hash"""
     block_hash = header.hash
     if block_hash in state.block_transactions:
         await _process_block_transactions(db_functions_async, state, header, chain)
 
 
-async def _process_block_transactions(db_functions_async: AsynchronousFunctions,
-        state: LateHeaderWorkerState, header: Header, chain: Chain) -> None:
-    assert app_state.headers is not None
-    longest_chain = cast(Chain, app_state.headers.longest_chain())
-    if chain is longest_chain:
-        confirmations = longest_chain.height - header.height
+async def _process_one_merkle_proof(db_functions_async: AsynchronousFunctions,
+        state: LateHeaderWorkerState, tx_hash: bytes, flags: TxFlags, tx_block_hash: bytes,
+        proof_data: bytes, confirmations: int, header: Header) -> None:
+    """raises `MerkleProofProcessingError` with an associated enum MerkleProofErrorType
+    enumerating the possible failure cases. These errors should be recoverable"""
+    # It is possible that the transaction may have changed from under us in the database,
+    # we need to check that it hasn't, but if it has skip the transaction.
+    if proof_data is None:
+        logger.error("Deferred verification transaction %s block hash now lacks proof data",
+            hash_to_hex_str(tx_hash))
+        raise MerkleProofProcessingError(MerkleProofErrorType.NO_PROOF_DATA)
```

Comment:
I do not know what raising these errors does, there is nothing we can do, we are logging the error case, we may as well return None.

Reply:
It was mostly to stay true to the original logic after extracting the function from the for loop...

I can return None but of these possible error cases do you want me to raise any of them?
Previously you were breaking out of the loop when encountering a bad server... Do you want an exception raised in that case?

Reply:
I will return None for all. Otherwise let me know.